### PR TITLE
(draft) feat: add Koala Backend API

### PR DESCRIPTION
**Pull request by stefan-cooper**
## Summary
This PR adds some basic infrastructure for a Koala API. The intention for this PR is purely to raise the discussion for how we implement this.

I propose we implement an API using [Quart](https://pgjones.gitlab.io/quart/) - Quart is forked from Flask but allows more async options. I had originally tried implementing via Flask, but ran into many issues when working with Discord.py.

I think the main point of contention and changes that will be required will be reducing and simplifying the usage of `ctx` (Discord Context) as that won't be available to the API. Where possible we should do things like `self.bot.get_guild(guild_id)` where the `guild_id` is a parameter on the methods (which can be passed as `ctx.guild.id`).

I'll add comments on other points which I think might be interesting and would be interested in hearing your opinions on.

<br>

- [x] All of this code is your own, or follows the author repo's licence.

Contributes to: #0 (Pending issue)

Signed-off-by: Stefan Cooper <stefan.cooper27@gmail.com>

### CODE REVIEW

File: koala/cogs/colour_role/cog.py [hunk 1]
```diff
@@ -253,8 +262,7 @@ async def prune_members_old_colour_roles(self, members: List[discord.Member]) ->
             return False
         return True
 
-    async def create_custom_colour_role(self, colour: discord.Colour, colour_str: str,
-                                        ctx: commands.Context) -> discord.Role:
+    async def create_custom_colour_role(self, colour: discord.Colour, colour_str: str, guild_id: int) -> discord.Role:
```
Comment (stefan-cooper) [inline koala/cogs/colour_role/cog.py:265]: Example of simplification of `ctx` -> `guild_id` as per PR description

File: koala/cogs/colour_role/cog.py [hunk 2]
```diff
@@ -72,6 +76,11 @@ def __init__(self, bot):
         self.bot = bot
         insert_extension("ColourRole", 0, True, True)
         self.cr_database_manager = ColourRoleDBManager()
+        colour_role_api.add_url_rule('/colour_role/create_role/<int:guild_id>/<string:colour_str>', view_func=self.create_role_api, methods=['POST'])
+
+    def create_role_api(self, guild_id: int, colour_str: str) -> discord.Role:
+        asyncio.run_coroutine_threadsafe(self.create_custom_colour_role(discord.Colour.random(), colour_str, guild_id), self.bot.loop).result()
+        return 200
```
Comment (stefan-cooper) [inline koala/cogs/colour_role/cog.py:83]: Example of how we would interact with existing methods in the system

File: koala/cogs/text_filter/cog.py [hunk 3]
```diff
@@ -47,6 +51,8 @@ def __init__(self, bot):
         self.bot = bot
         insert_extension("TextFilter", 0, True, True)
         self.tf_database_manager = TextFilterDBManager(bot)
+        
+        text_filter_api.add_url_rule('/text_filter/list_words/<int:guild_id>', view_func=self.tf_database_manager.get_filtered_text_for_guild, methods=['GET'])
```
Comment (stefan-cooper) [inline koala/cogs/text_filter/cog.py:55]: Example of interacting directly with the database for GETs

**Comment by JayDwee** [comment]
How does quart compare to aiohttp?

**Comment by stefan-cooper** [comment]
> How does quart compare to aiohttp?

Don't know much about `aiohttp` - i've been experimenting with different REST API options over the past week or so, Quart seemed the easiest to get up and running quickly but I think if I can get it working, then FastAPI would probably be the most well supported library to use.

**Comment by stefan-cooper** [comment]
Superseded by #299

**Pull request closed without merging**